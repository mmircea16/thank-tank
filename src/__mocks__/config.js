const config = jest.genMockFromModule('../config');
export default {gridSize: 1};