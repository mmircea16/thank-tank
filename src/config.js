export default {
    clientId: '', //google client ID
    scope: 'https://www.googleapis.com/auth/spreadsheets',
    spreadsheetId: '', //ID for the spreadsheet
    speedLowerBound: 25, //lower bound for flip randomisation
    speedUpperBound: 35, //upper bound for flip randomisation
    gridSize: 12, //grid size for the cards
    timePerWord: 0.1, //to work out how long to show a message
    range: 'A2:E', //range for the data columns in the spreadsheet
    mapping: {
        //data mapping columns from spreadsheet
        name: 1,
        from: 3,
        message: 2,
        secret: 4
    }
};
