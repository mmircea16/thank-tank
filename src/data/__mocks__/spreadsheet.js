const spreadSheet = jest.genMockFromModule('../spreadSheet');

export let checkAuth = spreadSheet.checkAuth;
export let loadData = spreadSheet.loadData;