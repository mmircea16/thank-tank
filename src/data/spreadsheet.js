import config from '../config';
/**
 * Get the user authentication status
 */
export function checkAuth(immediate) {
  return new Promise((resolve) => {

    window.gapi.load('client:auth2', () => {
      console.log("config", config)
      window.gapi.client.init({
        'clientId': config.clientId,
        'scope': config.scope,
        'discoveryDocs': ["https://sheets.googleapis.com/$discovery/rest?version=v4"]
      }).then( () => {
        console.log('resolve');
        resolve()
      });
    });
  });
}

/**
 * Load the quotes from the spreadsheet
 * Embellish them with user own likes
 */
export function loadData() {
  var promise = new Promise((resolve, reject) => {
    window.gapi.client.load('sheets', 'v4', () => {
      window.gapi.client.sheets.spreadsheets.values.batchGet({
        spreadsheetId: config.spreadsheetId,
        ranges: config.ranges,
      }).then(
        (response) => {
          let allRanges = response.result.valueRanges;
          let allThanks = allRanges.flatMap(range => range.values);
          resolve(filterEmptyThankees(allThanks));
        },
        (response) => {
          reject(false, response.result.error);
        }
      );
    });

  });

  return promise;
}

let filterEmptyThankees = function (values) {
  if (!values) return [];

  return values.filter(row => row[config.mapping.name] && (row[config.mapping.name] !== ''))
};