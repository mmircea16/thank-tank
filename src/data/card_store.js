"use srtict";
import { checkAuth, loadData } from './spreadsheet'
import config from '../config';

var mapping = config.mapping;
var actualDataPromise;
var isPending = false;
var lastShown = -1;

function fetchData() {
    if (!isPending) {

        isPending = true;
        actualDataPromise = new Promise((resolve) => {
            checkAuth(true).then((authResult) => {
                loadData()
                    .then(resolve)
                    .then(() => {
                        isPending = false;
                    });
            });
        });
    }
}

function getShowIndex(rowCount){
    var showIndex = lastShown + 1;
    if(showIndex < rowCount)
        return showIndex;
    return Math.floor(Math.random()*rowCount);
}

class CardStore {
    constructor() {
        fetchData();
    }

    getNext() {
        fetchData();
        return new Promise((resolve, reject) => {
            actualDataPromise.then((rows) => {
                var showIndex = getShowIndex(rows.length);
                var hasName = rows[showIndex][mapping.secret] !== 'Yes';
                var from = hasName && rows[showIndex][mapping.from] || 'Anonymous';
                resolve({
                    name: rows[showIndex][mapping.name],
                    message: rows[showIndex][mapping.message],
                    from: from
                });
                lastShown = showIndex > lastShown ? showIndex : lastShown;
            });
        });
    }
}

export function reset() {
    console.warn('depricated call reset (only allowed in tests)');
    lastShown = -1;
    isPending = false;
}

export default function getInstance() {
    return new CardStore();
}