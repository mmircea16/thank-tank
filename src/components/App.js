import React, { Component } from 'react';

import { checkAuth, loadData } from '../data/spreadsheet';
import FlippingCards from './FlippingCards';
import config from '../config'

class App extends Component {

  constructor(props) {
    super(props);

    this.state = {
      loading: true,
      error: null,
      hasContent: false,
      authenticated: false
    }

    this.onLoad = this.onLoad.bind(this);
    this.onError = this.onError.bind(this);
    this.updateSigninStatus = this.updateSigninStatus.bind(this);
  }

  componentDidMount() {
    checkAuth().then(this.handleAuth.bind(this));
  }

  /**
   * Check user authenification status and set app state accordingly
   */
  handleAuth() {
    window.gapi.auth2.getAuthInstance().isSignedIn.listen(this.updateSigninStatus)
    let signedIn = window.gapi.auth2.getAuthInstance().isSignedIn.get();
    console.log("handleAuth", signedIn)
    this.updateSigninStatus(signedIn);
  }

  updateSigninStatus(signedIn) {
    if (signedIn) {
      this.setState({
        authenticated: true
      });
      loadData().then(this.onLoad, this.onError);
    } else {
      this.setState({
        authenticated: false,
        loading: false
      })
    }
  }

  onLoad(data, error) {
    this.setState({
      rows: data,
      hasContent: true,
      loading: false
    });
  }

  onError(error) {
    this.setState({
      error: error,
      hasContent: false,
      loading: false
    });
  }

  render() {
    return (
      <div className="app">
        <div className="banner">
          <div className="left"></div>
          <div className="right"></div>
        </div>
        {this.renderContent()}
      </div>
    );
  }

  renderContent() {
    if (this.state.loading) {
      return (
        <div className="loader" />
      );
    }

    if (!this.state.authenticated) {
      return (
        <button onClick={this.authenticate.bind(this)} className="btn">Connect with Google</button>
      );
    }
    
    if (this.state.hasContent) {
      var pairs = new Array(config.gridSize).fill(1);
      var rows = pairs.map((row, index) => {
        return (<FlippingCards key={'flip-' + index} />)
      });

      return (<div className="cards">
        {rows}
      </div>
      );
    }

    if (this.state.error) {
      return (
        <h1> {this.state.error} </h1>
      );
    }
  }

  /**
   * Request Google authentification
   */
  authenticate(e) {
    e.preventDefault();
    console.log("authenticate")
    window.gapi.auth2.getAuthInstance().signIn();
    // checkAuth(false).then(this.handleAuth.bind(this));
  }
}

export default App;
