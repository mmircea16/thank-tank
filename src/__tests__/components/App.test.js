'use strict';

import React from 'react';
import renderer from 'react-test-renderer';
jest.mock('../../data/spreadsheet');
jest.mock('../../helpers/randomColour');
jest.mock('../../data/card_store');
jest.mock('../../config');
import { checkAuth, loadData } from '../../data/spreadsheet';
import getCardStore from '../../data/card_store';
import config from '../../config';
import App from '../../components/App';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

describe("App", () => {
    it('render if not authenticated', (done) => {
        var promise = new Promise((resolve, reject) => {
            var authPromise = new Promise((resolve, reject) => resolve(false))
            checkAuth.mockReturnValue(authPromise);
            var app = renderer.create(<App />);
            authPromise.then(resolve.bind(null, app), reject);
        });
        promise.then((app) => {
            expect(app.toJSON()).toMatchSnapshot();
            done();
        });
    });

    it('render if authenticated and waiting for data', (done) => {
        var promise = new Promise((resolve, reject) => {
            var authPromise = new Promise((resolve, reject) => resolve(true))
            checkAuth.mockReturnValue(authPromise);
            var app = renderer.create(<App />);
            authPromise.then(resolve.bind(null, app), reject);
        });
        promise.then((app) => {
            expect(app.toJSON()).toMatchSnapshot();
            done();
        });
    });


    it('render if authenticated and data arrived', (done) => {
        var promise = new Promise((resolve, reject) => {
            var app = renderer.create(<App />);
            var getNextPromise = new Promise((resolveNext) => {
                var loadDataPromise = new Promise((resolve, reject) => {
                    resolve([['time', 'email', 'name', 'message', 'No'], ['time', 'email', 'name', 'message', 'No']]);
                });
                var authPromise = new Promise((resolve, reject) => resolve(true))
                checkAuth.mockReturnValue(authPromise);
                loadData.mockReturnValue(loadDataPromise);
                loadDataPromise.then(resolveNext.bind(null, {name:'name', message:'message', from:'thanked By'}));
            });
            getCardStore.mockReturnValue({ getNext: () => getNextPromise });
            getNextPromise.then(resolve.bind(null, app), reject);
        });
        promise.then((app) => {
            expect(app.toJSON()).toMatchSnapshot();
            done();
        });
    });
});