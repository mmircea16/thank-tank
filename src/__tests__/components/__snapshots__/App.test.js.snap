// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App render if authenticated and data arrived 1`] = `
<div
  className="app"
>
  <div
    className="banner"
  >
    <div
      className="left"
    />
    <div
      className="right"
    />
  </div>
  <div
    className="cards"
  >
    <section
      className="container"
    >
      <div
        className="card false"
      >
        <article
          className="thankCard front blueCard"
        >
          <header>
            <strong>
              Thank you 
              <span
                className="name"
              >
                name
                !
              </span>
            </strong>
          </header>
          <div
            className="autoscrollcomponent"
          >
            <section>
              message
            </section>
          </div>
          <footer>
            from 
            thanked By
          </footer>
        </article>
        <article
          className="thankCard back blueCard"
        >
          <header>
            <strong>
              Thank you 
              <span
                className="name"
              >
                dummy
                !
              </span>
            </strong>
          </header>
          <div
            className="autoscrollcomponent"
          >
            <section>
              dummy
            </section>
          </div>
          <footer>
            from 
            dummy
          </footer>
        </article>
      </div>
    </section>
  </div>
</div>
`;

exports[`App render if authenticated and waiting for data 1`] = `
<div
  className="app"
>
  <div
    className="banner"
  >
    <div
      className="left"
    />
    <div
      className="right"
    />
  </div>
  <div
    className="loader"
  />
</div>
`;

exports[`App render if not authenticated 1`] = `
<div
  className="app"
>
  <div
    className="banner"
  >
    <div
      className="left"
    />
    <div
      className="right"
    />
  </div>
  <button
    className="btn"
    onClick={[Function]}
  >
    Connect with Google
  </button>
</div>
`;
